src/operations/map/MapOperations.ts

```diff
@@ -86,44 +107,97 @@ export class MapOperations {
+        baseMap.getCanvas().style.cursor = '';
+      }
+
+      if (!feature) {
+        popup.remove();
+        return;
+      }
+
+      baseMap.getCanvas().style.cursor = 'pointer';
+      hoveredPolygonId = feature.id;
+      if (hoveredPolygonId) {
+        baseMap.setFeatureState({ source: 'countries', id: hoveredPolygonId }, { hover: true });
+      }
+
+      switch (selectedMapType) {
+        case 'ipc': {
+          const countryName = feature.properties.adm0_name;
+          const ipcColorMap = IPCMapOperations.generateColorMap(ipcData!, countries!);
+
+          if (!ipcColorMap[countryName]) {
+            popup.remove();
+            return;
+          }
+
+          const { date_of_analysis, analysis_period, ipc_percent, ipc_popnmbr } = IPCMapOperations.findIpcData(
+            countryName,
+            ipcData!
+          );
+          const formattedPopNum = ipc_popnmbr.toLocaleString('en-US', {
+            useGrouping: true,
+          });
+          const title = `${countryName}`;
+          const body = `Date of analysis: ${date_of_analysis} <br> Validity period: ${analysis_period}`;
+          const conclusion = `${formattedPopNum} people in IPC/CH Phase 3 and above <br> (${ipc_percent}% of people in the analyzed areas)`;
+
+          const popupContent = this.generatePopupContent(title, body, conclusion);
+          popup.setLngLat(e.lngLat).setHTML(popupContent).addTo(baseMap);
+          break;
         }
+        default:
+          popup.remove();
+          return;
       }
     });
 
     baseMap.on('mouseleave', 'country-fills', () => {
+      baseMap.getCanvas().style.cursor = '';
       if (hoveredPolygonId) {
         baseMap.setFeatureState({ source: 'countries', id: hoveredPolygonId }, { hover: false });
+        hoveredPolygonId = undefined;
       }
-      hoveredPolygonId = undefined;
+      popup.remove();
     });
 
     let isDragging = false;
     baseMap.on('mousedown', () => {
       isDragging = false;
     });
-
     baseMap.on('mousemove', () => {
       isDragging = true;
     });
-
-    baseMap.on('mouseup', 'country-fills', (e) => {
-      if (!isDragging && e.features && (e.features[0] as unknown as CountryMapData).properties.interactive) {
-        alert(`You clicked on ${(e.features[0] as unknown as CountryMapData).properties.adm0_name}`);
+    baseMap.on('mouseup', 'country-fills', (e: any) => {
+      if (!isDragging && e.features && (e.features[0] as CountryMapData).properties.interactive) {
+        alert(`You clicked on ${(e.features[0] as CountryMapData).properties.adm0_name}`);
       }
     });
   }
 
+  static generatePopupContent(title?: string, body?: string, conclusion?: string): string {
+    return `
+      <div class="p-4 bg-white dark:bg-black text-black dark:text-white rounded-medium flex flex-col gap-2">
+        <h1 class="text-md font-semibold">${title}</h1>
+        <p class="text-xs">${body}</p>
+        <p class="text-xs">${conclusion}</p>
+      </div>
+    `;
+  }
```

Comment:
Maybe that's just a personal opinion, but I'd inline this function as it is used only once and does nothing complicated.

Reply:
That's the better way to do it like it already is, to separate the concerns. Maybe a precise naming here would be not bad

Reply:
done